File: 프로그래머스/3/81303. 표 편집/표 편집.py
class Node:
    def __init__(self, left=None, right=None):
        self.left = left
        self.right = right
        self.removed = False
    
    
class Table:
    def __init__(self, n, k):
        self.table = [Node(i - 1, i + 1) for i in range(n)]
        self.table[0].left = None
        self.table[-1].right = None
        self.removed = []
        self.k = k
        
    def up(self, x):
        for _ in range(x):
            self.k = self.table[self.k].left
    
    def down(self, x):
        for _ in range(x):
            self.k = self.table[self.k].right
    
    def delete(self):
        self.removed.append(self.k)
        self.table[self.k].removed = True
        left, right = self.table[self.k].left, self.table[self.k].right
        
        if left is not None:
            self.table[left].right = right
            self.k = left
        if right is not None:
            self.table[right].left = left
            self.k = right
    
    def restore(self):
        node = self.removed.pop()
        self.table[node].removed = False
        left, right = self.table[node].left, self.table[node].right
        
        if left is not None:
            self.table[left].right = node
        if right is not None:
            self.table[right].left = node
        
    def get_result(self):
        answer = ''
        for node in self.table:
            answer += 'X' if node.removed else 'O'
        return answer


def solution(n, k, cmd):
    t = Table(n, k)
    for target in cmd:
        if target[0] == 'C':
            t.delete()
        elif target[0] == 'Z':
            t.restore()
        elif target[0] == 'U':
            t.up(int(target[2:]))
        elif target[0] == 'D':
            t.down(int(target[2:]))
    
    return t.get_result()

File: 프로그래머스/3/81303. 표 편집/표 편집.java
import java.util.*;

class Solution {

    private int[] left;
    private int[] right;
    private Deque<Integer> removed;
    private int k;
    private int n;

    public String solution(int n, int k, String[] cmd) {
        this.n = n;
        this.k = k;
        left = new int[n];
        right = new int[n];
        removed = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            left[i] = i - 1;
            right[i] = i + 1;
        }

        for (String target : cmd) {
            char c = target.charAt(0);
            if (c == 'C') {
                delete();
            } else if (c == 'Z') {
                restore();
            } else {
                int x = Integer.parseInt(target.substring(2));
                if (c == 'U') up(x);
                else down(x);
            }
        }

        return getResult();
    }

    private void up(int x) {
        for (int i = 0; i < x; i++) k = left[k];
    }

    private void down(int x) {
        for (int i = 0; i < x; i++) k = right[k];
    }

    private void delete() {
        removed.push(k);
        int l = left[k];
        int r = right[k];
        if (l >= 0) right[l] = r;
        if (r < n) left[r] = l;
        k = (r < n) ? r : l;
    }

    private void restore() {
        int node = removed.pop();
        int l = left[node];
        int r = right[node];
        if (l >= 0) right[l] = node;
        if (r < n) left[r] = node;
    }

    private String getResult() {
        boolean[] isRemoved = new boolean[n];
        for (int node : removed) isRemoved[node] = true;

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(isRemoved[i] ? 'X' : 'O');
        }
        return sb.toString();
    }
}